### PR TITLE
Mimecast integration ciac 9090

### PR DESCRIPTION
**Pull request by YairGlik**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-9090

## Description
A few sentences describing the overall goals of the pull request's commits.

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 1]
```diff
@@ -2040,6 +2070,1035 @@ script:
...
+      defaultValue: yes
+    - name: enforced
+      description: Determines if the policy enforcement option is enabled.
+      auto: PREDEFINED
+      predefined:
+      - yes
+      - no
+      defaultValue: no
+    - name: from_date
+      description: string (yyyy-MM-dd'T'HH:mm:ssZ) Specifies the date of a policy should go into effect, in ISO 8601 format (e.g. 2015-11-16T14:49:18+0000). When specified, this will override the fromEternal value to false.
```
Comment (jbabazadeh) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:2866]: should also have the option to relative time like '3 days ago'

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 2]
```diff
@@ -2040,6 +2070,1035 @@ script:
...
+      required: true
+    - name: enabled
+      description: Determines if the policy should be enabled to process messages. The default value is true.
+      auto: PREDEFINED
+      predefined:
+      - yes
+      - no
+      defaultValue: yes
+    - name: from_date
+      description: (yyyy-MM-dd'T'HH:mm:ssZ) Specifies the date of a policy should go into effect, in ISO 8601 format (e.g. 2015-11-16T14:49:18+0000). When specified, this will override the fromEternal value to false.
```
Comment (jbabazadeh) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:2734]: should also have the option to provide relative time.

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 3]
```diff
@@ -2040,6 +2070,1035 @@ script:
...
+      - no
+      defaultValue: no
+    - name: from_attribute_id
+      description: The secure id of the address attribute.
+    - name: from_attribute_name
+      description: Name of address attribute.
+    - name: from_attribute_value
+      description: Value to which address attribute is equal to.
+    - name: from_date
+      description: Specifies the date of a policy should go into effect, in ISO 8601 format (e.g. 2015-11-16T14:49:18+0000). When specified, this will override the fromEternal value to false.
```
Comment (jbabazadeh) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:2602]: should have relative time?

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2_description.md [hunk 4]
```diff
@@ -1,12 +1,24 @@
+### API 2.0 Settings
```
Comment (jbabazadeh) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2_description.md:1]:
```suggestion
### API 2.0 Authentication
```

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2_description.md [hunk 5]
```diff
@@ -1,12 +1,24 @@
...
+To use API 2.0, you need to generate a **Client ID** and **Client Secret** from the Mimecast admin console:
+
+1. **Navigate to:** Admin | Services | API and Platform Integrations | Available integrations.
+2. Locate the Mimecast API 2.0 tile and select **Generate Keys**.
+
+Make sure you have provided the Client ID and Client Secret, checked the set the BaseUrl to `https://api.services.mimecast.com`.
+
+
+### API 1.0 Settings
```
Comment (jbabazadeh) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2_description.md:10]:
```suggestion
### API 1.0 Authentication
```

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2_description.md [hunk 6]
```diff
@@ -1,12 +1,24 @@
...
 mimecast-refresh-token -> refreshes the validity duration of access key & secret key (3 days)
 
-2. In order to use the rest of the commands, you are required to provide: App ID, App Key, Access Key & Secret Key.
+1. In order to use the rest of the commands, you are required to provide: App ID, App Key, Access Key & Secret Key.
 For detailed information about creating these fields, please refer to the [Mimecast Documentation](https://integrations.mimecast.com/documentation/api-overview/authentication-scripts-server-apps/).
 
-3. Fetch Incidents - the integration has the ability to fetch 3 types of incidents: url, attachment & impersonation.
+
+### Fetch Incidents
+the integration has the ability to fetch 3 types of incidents: url, attachment & impersonation.
```
Comment (jbabazadeh) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2_description.md:23]: add a rate limit section.

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2_test.py [hunk 7]
```diff
@@ -686,3 +686,275 @@ def test_get_message_metadata_with_attachments(mocker):
...
+
+def test_list_account_command(mocker):
+    """
+
+
+    Args:
+    """
+
+    args = {'account_code': 'ABC123'}
+    expected_response = {'meta': {'status': 200}, 'data': [{'region': 'us', 'archive': False, 'gateway': True, 'passphrase': '', 'supportCode': '6747', 'maxRetention': 30, 'maxRetentionConfirmed': True, 'minRetentionEnabled': False, 'automatedSegmentPurge': True, 'type': 'full', 'policyInheritance': False, 'databaseCode': 'usterm13', 'searchReason': False, 'contentAdministratorDefaultView': 'Metadata', 'adminSessionTimeout': 60, 'exportApi': False, 'exgestAllowQuery': False, 'exgestAllowExtraction': True, 'expressAccount': False, 'cybergraphV2Enabled': False, 'accountCode': 'ABC123', 'accountName': 'API Alliance - Palo Alto Networks', 'adminEmail': '', 'contactEmail': 'techbd@paloaltonetworks.mime.integration.com', 'domain': '', 'userCount': 10, 'mimecastId': '01-0102-00236', 'contactName': 'Adnan Kharuf', 'telephone': '4088307584', 'packages': ['Journal Services [1053]',
```
Comment (jbabazadeh) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2_test.py:834]: move all the response to a different file

File: Packs/Mimecast/Integrations/MimecastV2/README.md [hunk 8]
```diff
@@ -9,8 +9,8 @@ Please read detailed instructions in order to understand how to set the integrat
```
Comment (jbabazadeh) [inline Packs/Mimecast/Integrations/MimecastV2/README.md:9]: you should add -

1. The new command
2. Update the deprecate commands
3. Instruction about the api 2.0
4. Rate limit section.

**Review by jbabazadeh — CHANGES_REQUESTED** [review]
Great work!! ⭐ 
Please see my comments
you should add more unit test in order to have enough coverage

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 9]
```diff
@@ -17,12 +17,17 @@
 
 
 ''' GLOBALS/PARAMS '''
-
 BASE_URL = demisto.params().get('baseUrl')
 ACCESS_KEY = demisto.params().get('accessKey')
-SECRET_KEY = demisto.params().get('secretKey') or demisto.params().get('secretKey_creds', {}).get('password', '')
+
+SECRET_KEY = demisto.params().get('secretKey') or demisto.params().get('secretKey_creds')
```
[the remark this answers is not in the document]
Comment (YairGlik) [reply Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:23]: When using api 2.0 then secretKey_creds is an empty string, so get('password', '') get an error

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 10]
```diff
@@ -33,12 +38,23 @@
...
+EMAIL_ADDRESS = demisto.params().get('email') or demisto.params().get('credentials', {})
+if isinstance(EMAIL_ADDRESS, dict):
+    EMAIL_ADDRESS = EMAIL_ADDRESS.get('identifier', '')
+PASSWORD = demisto.params().get('password') or demisto.params().get('credentials', {})
+if isinstance(PASSWORD, dict):
+    PASSWORD = PASSWORD.get('password', '')
 FETCH_DELTA = int(demisto.params().get('fetchDelta', 24))
 
 
+CLIENT_ID = demisto.params().get('client_id')
```
[the remark this answers is not in the document]
Comment (YairGlik) [reply Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:50]: I added

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 11]
```diff
@@ -1,54 +1,59 @@
...
+- display: Client ID (for API 2.0)
+  name: client_id
+  type: 4
+- display: Client Secret (for API 2.0)
+  name: client_secret
+  type: 4
+- display: App ID (for API 1.0)
   name: appId
-  required: true
   type: 0
```
[the remark this answers is not in the document]
Comment (YairGlik) [reply Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:19]: you mean section: Connect?

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 12]
```diff
@@ -1,54 +1,59 @@
...
-- display: User Email Address (Use for auto token refresh)
+- display: User Email Address (Use for auto token refresh) (for API 1.0)
   name: credentials
   type: 9
-  displaypassword: Password
+  displaypassword: Password (for API 1.0)
   required: false
-- display: User Email Address (Use for auto token refresh)
+- display: User Email Address (Use for auto token refresh) (for API 1.0)
   name: email
```
[the remark this answers is not in the document]
Comment (YairGlik) [reply Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:26]: you mean section: Connect?

File: Packs/Mimecast/ReleaseNotes/2_3_0.md [hunk 13]
```diff
@@ -0,0 +1,22 @@
+#### Integrations
+
+##### Mimecast v2
+
+- Added support for **API 2.0** with **client id** and **secret id** parameters.
+- Deprecated the command ***mimecast-update-policy***. Use ***mimecast-update-block-sender-policy*** instead.
+- Deprecated the command ***mimecast-create-policy***. Use ***mimecast-create-block-sender-policy*** instead.
+- Deprecated the command ***mimecast-list-blocked-sender-policies***. Use ***mimecast-list-policies*** instead.
+- Added support for additional types of policies in the command ***mimecast-get-policy***.
+- Added support for additional types of policies in the command ***mimecast-delete-policy***.
```
Comment (ShirleyDenkberg) [inline Packs/Mimecast/ReleaseNotes/2_3_0.md:10]:
```suggestion
- Added support for additional types of policies in the ***mimecast-get-policy*** command .
- Added support for additional types of policies in the ***mimecast-delete-policy*** command .
```

File: Packs/Mimecast/ReleaseNotes/2_3_0.md [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
```
Comment (ShirleyDenkberg) [inline Packs/Mimecast/ReleaseNotes/2_3_0.md:11]: - Add the following commands:

File: Packs/Mimecast/ReleaseNotes/2_3_0.md [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+- ***mimecast-get-search-logs***: Retrieves the search logs.
+- ***mimecast-get-view-logs***: Retrieves the email view logs.
+- ***mimecast-list-account***: Retrieves the summary information for a Mimecast account.
+- ***mimecast-list-policies***: Retrieves a list of policies of a specified type.
+- ***mimecast-create-block-sender-policy***: Creates a block-sender policy.
+- ***mimecast-update-block-sender-policy***: Updates a block-sender policy.
+- ***mimecast-create-antispoofing-bypass-policy***: Creates an anti spoofing bypass policy.
+- ***mimecast-update-antispoofing-bypass-policy***: Updates an anti spoofing bypass policy.
+- ***mimecast-create-address-alteration-policy***: Creates an address alteration policy.
+- ***mimecast-update-address-alteration-policy***: Updates an address alteration policy.
```
Comment (ShirleyDenkberg) [inline Packs/Mimecast/ReleaseNotes/2_3_0.md:22]:
```suggestion
   - ***mimecast-get-archive-search-logs***: Retrieves archive search logs.
   - ***mimecast-get-search-logs***: Retrieves the search logs.
   - ***mimecast-get-view-logs***: Retrieves the email view logs.
   - ***mimecast-list-account***: Retrieves the summary information for a Mimecast account.
   - ***mimecast-list-policies***: Retrieves a list of policies of a specified type.
   - ***mimecast-create-block-sender-policy***: Creates a block-sender policy.
   - ***mimecast-update-block-sender-policy***: Updates a block-sender policy.
   - ***mimecast-create-antispoofing-bypass-policy***: Creates an anti spoofing bypass policy.
   - ***mimecast-update-antispoofing-bypass-policy***: Updates an anti spoofing bypass policy.
   - ***mimecast-create-address-alteration-policy***: Creates an address alteration policy.
   - ***mimecast-update-address-alteration-policy***: Updates an address alteration policy.
```

**Comment by ShirleyDenkberg** [comment]
@YairGlik After implementing my comments in the yml file, please regenerate the readme.
@jbabazadeh Doc review completed.

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 16]
```diff
@@ -2040,12 +2099,1049 @@ script:
...
+      required: true
+    - name: enabled
+      description: Determines if the policy should be enabled to process messages. The default value is true.
+      auto: PREDEFINED
+      predefined:
+      - yes
+      - no
+      defaultValue: yes
+    - name: from_date
+      description: (yyyy-MM-dd'T'HH:mm:ssZ) Specifies the date of a policy should go into effect, in ISO 8601 format (e.g. 2015-11-16T14:49:18+0000). When specified, this will override the fromEternal value to false.
```
[the remark this answers is not in the document]
Comment (YairGlik) [reply Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:2767]: I want to add:
or a relative time such as "3 days"
It's OK ?

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 17]
```diff
@@ -368,9 +418,17 @@ script:
       description: The receiver type.
       type: String
   - arguments:
-    - description: Policy ID.
+    - description: Policy ID. The policy ID can be retrieved using mimecast-list-policies use the data.id field.
```
[the remark this answers is not in the document]
Comment (YairGlik) [reply Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:421]: No. To delete a policy you always need Policy ID and not addressAlterationSetId

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml (hunk 16 above)
[the remark this answers is not in the document]
Comment (ShirleyDenkberg) [reply Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:2767]: yes

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 18]
```diff
@@ -2988,8 +3154,426 @@ def list_email_queues_command(args):
...
+    page = arg_to_number(args.get('page'))
+    page_size = arg_to_number(args.get('page_size'))
+    limit = arg_to_number(args.get('limit'))
+
+    data = [{}]
+
+    if query:
+        data[0]['query'] = query
+
+    result_list = request_with_pagination_api2('/api/archive/get-archive-search-logs',
```
[the remark this answers is not in the document]
Comment (YairGlik) [reply Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py]: This command can run in api 2.0 and also in api 1.0

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 19]
```diff
@@ -2040,12 +2099,1049 @@ script:
...
+      description: Number of items per page for pagination. Default value is '50'.
+    - name: limit
+      description: The maximum number of items to returned.
+      defaultValue: 50
+    outputs:
+    - contextPath: Mimecast.Account.region
+      description: 'The region where the account is hosted.'
+      type: String
+    - contextPath: Mimecast.Account.archive
+      description: 'If gateway features are enabled on the account.'
```
[the remark this answers is not in the document]
Comment (YairGlik) [reply Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:2228]: You're right, I fixed it

File: Packs/Mimecast/ReleaseNotes/2_3_0.md (hunk 14 above)
Comment (YairGlik) [reply Packs/Mimecast/ReleaseNotes/2_3_0.md]: It falls in Bild:
"is not using one of our templates, consider changing it to fit our standard."

is that ok?
Updated the integration to add the following commands:

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 20]
```diff
@@ -2596,45 +2596,195 @@ script:
       description: The maximum number of results to return.
       defaultValue: 50
     outputs:
-    - contextPath: Mimecast.Policy.ID
+    - contextPath: Mimecast.AddressAlterationPolicy.ID
```
Comment (jbabazadeh) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:2599]: is it a TPM request?

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 21]
```diff
@@ -893,13 +961,88 @@ def get_policy():
...
-    # Setup required variables
-    api_endpoint = '/api/policy/blockedsenders/get-policy'
+def get_policy_command(args):
+    headers = ['Policy ID', 'Sender', 'Reciever', 'Bidirectional', 'Start', 'End']
+    contents = []
+    policy_id = args.get('policyID')
+    policy_type = args.get('policyType', 'blockedsenders')
+    title = f'Mimecast list {policy_type} policies: \n These are the existing {policy_type} Policies:'
+
+    title = f'Mimecast Get {policy_type} Policy'
```
Comment (jbabazadeh) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:971]: remove one.

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 22]
```diff
@@ -244,40 +286,108 @@ script:
...
+      defaultValue: blockedsenders
+      auto: PREDEFINED
+      predefined:
+      - blockedsenders
+      - antispoofing-bypass
+      - address-alteration
+    description: Get a policy by ID.
     name: mimecast-get-policy
     outputs:
-    - contextPath: Mimecast.Policy.ID
```
Comment (jbabazadeh) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:253]: is it requested by the TPM?

**Review by jbabazadeh — CHANGES_REQUESTED** [review]
Please see my comments. and lets talk about the test playbook.

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml (hunk 20 above)
Comment (YairGlik) [reply Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:2599]: yes

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 23]
```diff
@@ -33,12 +38,25 @@
...
+PASSWORD = demisto.params().get('password') or demisto.params().get('credentials', {})
+if isinstance(PASSWORD, dict):
+    PASSWORD = PASSWORD.get('password', '')
 FETCH_DELTA = int(demisto.params().get('fetchDelta', 24))
 
 
+CLIENT_ID = demisto.params().get('client_id')
+CLIENT_SECRET = demisto.params().get('client_secret', {}).get("password") if demisto.params().get('client_secret') else None
+USE_OAUTH2 = bool(CLIENT_ID and CLIENT_SECRET)
+demisto.info(f"{USE_OAUTH2=}")
```
Comment (jbabazadeh) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:53]: remove print

File: Tests/conf.json [hunk 24]
```diff
@@ -1195,7 +1195,13 @@
         },
         {
             "integrations": "MimecastV2",
-            "playbookID": "Mimecast test"
+            "playbookID": "Mimecast test",
+            "instance_names": "api1"
```
Comment (jbabazadeh) [inline Tests/conf.json:1199]: can we change it to -
```suggestion
            "instance_names": "mimecast_api_v1"
```

File: Tests/conf.json [hunk 25]
```diff
@@ -1195,7 +1195,13 @@
...
+        },
+        {
+            "integrations": "MimecastV2",
+            "playbookID": "Mimecast Test api 2.0",
+            "instance_names": "api2"
```
Comment (jbabazadeh) [inline Tests/conf.json:1204]:
```suggestion
            "instance_names": "mimecast_api_v2"
```

**Review by jbabazadeh — APPROVED** [review]
just a small comments.

**Pull request merged**